packages/api/amplify_api/lib/src/graphql/ws/websocket_connection.dart

```diff
@@ -0,0 +1,186 @@
+
+  /// Fires when the connection is ready to be listened to, i.e.
+  /// after the first `connection_ack` message.
+  Future<void> get ready => _connectionReady.future;
+
+  /// Re-broadcasts messages for child streams.
+  final StreamController<WebSocketMessage> _rebroadcastController =
+      StreamController<WebSocketMessage>.broadcast();
+
+  /// Incoming message stream for all events.
+  Stream<WebSocketMessage> get _messageStream => _rebroadcastController.stream;
+
+  /// {@macro websocket_connection}
+  WebSocketConnection(this._config, this.authProviderRepo);
+
+  /// Connects to the real time WebSocket.
+  Future<void> _connect() async {
+    // Generate a URI for the connection and all subscriptions.
+    // See https://docs.aws.amazon.com/appsync/latest/devguide/real-time-websocket-client.html#handshake-details-to-establish-the-websocket-connection
+    final connectionMessage = WebSocketConnectionInitMessage(_config);
+    final authorizedConnectionMessage = await authorizeWebSocketMessage(
+            connectionMessage, _config, authProviderRepo)
+        as WebSocketConnectionInitMessage;
+    final connectionUri = authorizedConnectionMessage.getConnectionUri();
+
+    _channel = WebSocketChannel.connect(
+      connectionUri,
+      protocols: webSocketProtocols,
+    );
+    _subscription = _channel.stream
+        .transform(const WebSocketMessageStreamTransformer())
+        .listen(_onData);
+  }
+
+  /// Closes the WebSocket connection.
+  @override
+  void close() {
+    _subscription.cancel();
+    _channel.sink.close();
+  }
+
+  /// Initializes the connection.
+  Future<void> init() {
+    return _initFuture ??= _init();
+  }
+
+  Future<void> _init() async {
+    await _connect();
+    if (_connectionReady.isCompleted) return;
+    send(MessageType.connectionInit);
+    return ready;
+  }
+
+  /// Subscribes to the given GraphQL request. Returns the subscription object,
+  /// or throws an [Exception] if there's an error.
+  Stream<GraphQLResponse<T>> subscribe<T>(
+    GraphQLRequest<T> request,
+    void Function()? onEstablished,
+  ) {
+    if (!_connectionReady.isCompleted) {
+      init();
+    }
+    final subRegistration = WebSocketMessage(
+      messageType: MessageType.start,
+      payload:
+          SubscriptionRegistrationPayload(request: request, config: _config),
+    );
+    final subscriptionId = subRegistration.id!;
+    return _messageStream
+        .where((msg) => msg.id == subscriptionId)
+        .transform(
+            WebSocketSubscriptionStreamTransformer(request, onEstablished))
+        .asBroadcastStream(
+          onListen: (_) => _send(subRegistration),
+          onCancel: (_) => _cancel(subscriptionId),
+        );
+  }
+
+  /// Cancels a subscription.
+  void _cancel(String subscriptionId) {
+    _send(WebSocketMessage(
+      id: subscriptionId,
+      messageType: MessageType.stop,
+    ));
+    // TODO(equartey): if this is the only susbscription, close the connection.
```

Comment:
Is this how other clients do it? I guess it could make sense to save on pings but idk